compiler/rustc_typeck/src/collect/type_of.rs

```diff
@@ -497,7 +497,13 @@ pub(super) fn type_of(tcx: TyCtxt<'_>, def_id: DefId) -> Ty<'_> {
                     DUMMY_SP,
                     &format!("unexpected const parent in type_of(): {:?}", x),
                 ),
-            }
+            };
+
+            // Typeck doesn't expect erased regions to be returned from `type_of`.
+            tcx.fold_regions(ty, &mut false, |r, _| match r {
+                ty::ReErased => tcx.lifetimes.re_static,
+                _ => r,
+            })
```

Comment:
This doesn't seem right... once we erased regions, we shouldn't un-erase them. This can go wrong way too easily.

Reply:
The code above gets the type from `tcx.typeck(def_id).node_type(hir_id)`. The type retrieved has their lifetimes erased by `rustc_typeck::check::writeback` already.